Experimentation troubleshooting guide (#34682)
DataDog/documentation: content/en/product_analytics/experimentation/troubleshooting.md

   {{< img src="/product_analytics/experiment/troubleshooting_tooltip.png" alt="An experiment scorecard tooltip showing the metric name, the average user-level metric value per variant, the total metric value, and the user assignment count for each variant." style="width:90%;" >}}

1. **If a metric value is zero**, hover over the metric name and click the &#8942; menu item. Select **Edit Metric** to open the metric definition page.

1. Verify that the metric event name is correct (for example, check for typos). Check the event volume chart on the right side of the page to confirm the event is firing.

   {{< img src="/product_analytics/experiment/troubleshooting_metric_page.png" alt="The Edit Metric page showing the metric definition on the left and a bar chart of metric event volume over the past week on the right." style="width:90%;" >}}

   <div class="alert alert-info">A metric event must meet two criteria for Datadog to include it in experiment results:
   <ol>
   <li>The event must come from a user with at least one experiment exposure event.</li>
   <li>The event must occur after the user's first experiment exposure.</li>

>    </ol>

Reviewer: ```suggestion
   </ul>
```

{{% /collapse-content %}}

{{% collapse-content title="Verify subject key matching" level="h4" expanded=false id="verify-subject-key-matching" %}}

Datadog matches metric events to experiment exposures using a subject key. If the subject key in your SDK does not match the subject type attribute configured in Datadog, the experiment cannot associate metrics with users.

1. Under the **Flag & Exposures** tab, click **View Exposure Logs** to see a list of recently exposed subjects. For details on how exposure events are tracked, see the [SDK documentation][8].

   {{< img src="/product_analytics/experiment/troubleshooting_exposure_log1.png" alt="The Exposures log table showing columns for variant, timestamp, subject, flag key, experiment key, and allocation key for recently exposed users." style="width:90%;" >}}

1. Compare the value in the **Subject** column to the subject type attribute (typically `@usr.id`). The **Subject** column shows the value your SDK passes as [`targetingKey`][7].

> 1. If the values do not match, update either the [`targetingKey`][7] in your SDK or the subject type in Datadog to align them. You can view and configure subject types on the [subject types page][9]. By default, the subject type is `@usr.id`.

Reviewer 2: Clarified this a bit more, let me know if the new content makes sense!

**If the subject values match but experiment results are still missing**, continue to the next section to inspect individual sessions.

{{% /collapse-content %}}

{{% collapse-content title="Inspect individual sessions" level="h4" expanded=false id="inspect-individual-sessions" %}}

If subject values match and users are assigned to the experiment, inspect individual sessions to identify why specific users are not generating metric events.

1. On the [event stream page][4], filter sessions from the experiment by adding the following filter:

   ```
   @feature_flags.<flag-key>:<variant-value>
   ```

   {{< img src="/product_analytics/experiment/troubleshooting_event_stream.png" alt="The Product Analytics event stream filtered by @feature_flags.new-product-photos:false, showing a list of sessions with columns for date, session type, time spent, view count, error count, and action count." style="width:90%;" >}}

> 1. Select a session from a user assigned to the experiment. In the session timeline, check for two things:

Reviewer: ```suggestion
1. Select a session from a user assigned to the experiment. In the session timeline, check for the following:
```

   - **Is the metric event present?** Verify that the expected metric event is firing within the session.

>    - **Does the metric event occur after the feature flag evaluation?** Events that occur **before** the feature flag evaluates do not count toward experiment results.

Reviewer 2: @lukasgoetzweiss looping you on this comment from Jeff. Thanks in advance!

{{% collapse-content title="Check outlier handling" level="h4" expanded=false id="check-outlier-handling" %}}

If you have confirmed that metric events are firing and subject keys match, but metric values are still zero, outlier handling may be the cause.

When outlier handling is enabled, Datadog calculates a threshold based on the distribution of metric values across users. If the number of users with a metric event is small, Datadog may compute the threshold as zero, which truncates all metric values to zero.

To check if outlier handling is the cause:

1. On the [Experiments][2] page, hover over the metric name to click the &#8942; menu item. Select **Edit Metric** to open the metric definition page.
1. In the metric definition, toggle **Outlier handling** off.
1. Save the metric. To trigger an immediate recompute, click the &#8942; menu next to **Last Updated** in the **Metrics** section and click **run an update now**. Otherwise, wait for the next scheduled update.

{{< img src="/product_analytics/experiment/troubleshooting_recompute1.png" alt="The experiment page Metrics section showing the Last Updated menu with the option to run an update now." style="width:90%;" >}}

> If metric values appear after disabling outlier handling, the threshold was truncating your data. To resolve this, keep outlier handling disabled or set a higher threshold on the metric edit page.

Reviewer: ```suggestion
If metric values appear after disabling outlier handling, the threshold was truncating your data. To resolve this, keep outlier handling disabled or set a higher threshold on the Edit Metric page.
```

---
title: Troubleshooting
description: Troubleshoot issues when running experiments.
further_reading:
- link: "https://www.datadoghq.com/blog/datadog-product-analytics"
  tag: "Blog"
  text: "Make data-driven design decisions with Product Analytics"
---

## Overview

If you experience issues setting up or running Datadog Experiments, use this page to troubleshoot. If you continue to have trouble, [contact Datadog support][1].

## Experiment results do not appear

If experiment results are missing after you launch an experiment, start by checking whether the experiment is assigning users. Then, navigate to the appropriate troubleshooting step.

### Step 1: Confirm the experiment is assigning users

On the [Experiments][2] page, select your experiment and hover over the metric scorecard:

> - If the **User Assignment Count** for each variant is zero, start with [Step 2](#step-2-confirm-the-experiment-is-receiving-traffic) to debug traffic.

Reviewer: Looking at some of the experiments live on the site, i see that some of the Metrics say "Organization Assignment Count" instead of "User Assignment Count," and some say "Anonymous user Assignment Count". 

Not sure if this is also the case live in prod or this is just a byproduct of being on the preprod org. But if this field can differ by some setting or type of metric, is this relevant to users of this page? Do these instructions apply for all of those options?

- If you see assignments but no metric values, skip to [Step 3](#step-3-confirm-metric-events-are-firing).

> In the following example, the **User Assignment Count** is 12,427 for Variant A and 12,573 for Variant B.

Reviewer: ```suggestion

```

You don't really need to include this, the screenshot is clear and it doesn't add much value otherwise.

   {{< img src="/product_analytics/experiment/troubleshooting_tooltip.png" alt="An experiment scorecard tooltip showing the metric name, the average user-level metric value per variant, the total metric value, and the user assignment count for each variant." style="width:90%;" >}}

### Step 2: Confirm the experiment is receiving traffic

Verify that your feature flag is enabled, evaluates in the correct environment, and that traffic reaches the experiment's targeting rule.

> 1. On the experiment details page, hover over the experiment flag label (for example, `new-product-photos`).

Reviewer: They've gotten to this page via the previous step, so maybe introduce the naming there? 

I think it's fine to say "select your experiment to open its detail page." We don't actually use the title "details" anywhere on the page, so don't highlight it too much as the actual name of a page.

   {{< img src="/product_analytics/experiment/troubleshooting_feature_flag_page.png" alt="The Feature Flags page with the Enabled toggle highlighted in the top-right corner." style="width:90%;" >}}

1. In the **Real-time metric overview** section, confirm that the bar chart shows exposure events.

   {{< img src="/product_analytics/experiment/troubleshooting_flag_traffic1.png" alt="The Feature Flags page with the Environment dropdown highlighted in the Real-time metric overview section, showing a bar chart of exposures over time broken down by variant." style="width:90%;" >}}

Based on what you see in the **Real-time metric overview**, follow the appropriate path:

#### The flag is not receiving traffic

Confirm the flag is enabled in the correct environment. You can manage environments on the [Environments page][3].

See the [Getting Started with Feature Flags][5] guide for details on environments.

#### The flag is receiving traffic but experiment assignments are zero

If the flag shows exposures but the metric scorecard shows zero assignments, traffic is not reaching the experiment's [targeting rule][6].

> The **Targeting Rules & Rollouts** section displays a waterfall, a list of targeting rules that the flag evaluates from top to bottom. Rules above the experiment's targeting rule, such as rules that exclude internal users or specific organizations, can capture traffic before it reaches the experiment.

Reviewer: ```suggestion
The **Targeting Rules & Rollouts** section displays a list of targeting rules that the flag evaluates from top to bottom. Rules above the experiment's targeting rule, such as rules that exclude internal users or specific organizations, can capture traffic before it reaches the experiment.
```

   {{< img src="/product_analytics/experiment/troubleshooting_feature_flag_page.png" alt="The Feature Flags page with the Enabled toggle highlighted in the Targeting rules and rollouts section." style="width:90%;" >}}

1. In the **Real-time metric overview** section, confirm that the bar chart shows exposure events.

   {{< img src="/product_analytics/experiment/troubleshooting_flag_traffic1.png" alt="The Real-time metric overview section of the Feature Flags page, showing a bar chart of exposures over time broken down by variant." style="width:90%;" >}}

Based on what you see in the **Real-time metric overview**, follow the appropriate path:

#### The flag is not receiving traffic

If the **Real-time metric overview** section shows no exposure events, the flag is not receiving traffic from your application.

Confirm the flag is enabled in the environment where your application runs. You can manage environments on the [Environments page][3]. See the [Getting Started with Feature Flags][5] guide for details on environments.

> After you fix the flag configuration, check the **Real-time metric overview** for incoming exposure events. Then, return to [Step 1](#step-1-confirm-the-experiment-is-assigning-users) to verify that the subject assignment count is increasing.

Reviewer: ```suggestion
After you enable the flag, check the **Real-time metric overview** for incoming exposure events. Then, return to [Step 1](#step-1-confirm-the-experiment-is-assigning-users) to verify that the subject assignment count is increasing.
```

  tag: "Documentation"
  text: "Reading Experiment Results"
---

## Overview

If you experience issues setting up or configuring Datadog Experiments, use this page to start troubleshooting. If you continue to have trouble, [contact Datadog Support][1].

### Why are experiment results not showing up?

There are several reasons why experiment results might not show up when you first launch an experiment, this guide gives a step-by-step process for spotting and resolving issues.

#### Why am I not seeing any experiment exposures?
The first thing to check is that your experiment is actually being evaluated. To start, navigate to the flag page by clicking “Go to Flag”. This modal will also tell you which environment the experiment is associated with:

{{< img src="/product_analytics/experiment/troubleshooting_flag_link.png" alt="Link to flag" style="width:90%;" >}}

On this page you can see how much traffic is hitting your flag in real time. Make sure you are looking at the same environment that your experiment is associated with.

{{< img src="/product_analytics/experiment/troubleshooting_flag_traffic.png" alt="Flag traffic chart" style="width:90%;" >}}

> If you’re not seeing any exposures, make sure that the flag is enabled in the appropriate environment. You can manage environments on the [environments page](https://app.datadoghq.com/feature-flags/settings/environments).

Reviewer 3: Two issues:
1. Contraction: "If you're not seeing" → "If you are not seeing"
2. "make sure" → "confirm": "make sure that the flag is enabled" → "confirm the flag is enabled"

> If the flag is indeed being evaluated in the appropriate environment, confirm that traffic is hitting the experiment targeting rule within the waterfall:

Reviewer 3: Remove filler word "indeed": "If the flag is indeed being evaluated" → "If the flag is being evaluated"

{{< img src="/product_analytics/experiment/troubleshooting_flag_waterfall.png" alt="Flag waterfall" style="width:90%;" >}}

If you are not seeing any traffic going to the experiment’s targeting rule, check the following:

> 1. Are there targeting rules above the experiment that are removing all of the experiment’s traffic?

Reviewer 3: The numbered list items are formatted as questions, which is unusual. Rewrite as imperative instructions:
1. Check whether targeting rules above the experiment are removing all traffic.
2. Verify the filters in the experiment targeting rule are being met.
3. Check the traffic allocation setting.

2. Are the filters in the experiment targeting rule being met?
3. What is the traffic allocation set to?

> #### Why are my metric values all zero?

Reviewer 3: Heading as a question — suggest: `#### Why are my metric values all zero?` → `#### Metric values showing as zero`
